docs/howto/use_modules_and_layers.md

```diff
@@ -0,0 +1,392 @@
+
+```scala mdoc:silent
+// get back the repo from the mixed value:
+val log = mix.get[Has[Logger.Service]].log("Hello modules!")
+```
+
+As per the recipe above, it is extremely convenient to define a type alias for `Has[Service]`.
+Usually we don't create a `Has` directly, but we do that through `ZLayer`.
+
+### The `ZLayer` data type
+
+`ZLayer[-RIn, +E, +ROut <: Has[_]]` is a recipe to build an environment of type `ROut`, starting from a value `RIn`, possibly producing an error `E` during creation. 
+
+In adherence with environmental concepts, the absence of a required input is represented by `RIn = Any`, conveniently  used in the type alias `ZLayer#NoDeps`.
+
+The `ZLayer` companion object offers a number of constructors to build layers (from pure values, from managed resources to guarantee acquire/release, from other services). 
+
+We can compose `layerA` and `layerB`  _horizontally_ to build a layer that has the requirements of both layers, to provide the capabilities of both layers, through `layerA ++ layerB`
+ 
+We can also compose layers _vertically_, meaning the output of one layer is used as input for the subsequent layer to build the next layer, resulting in one layer with the requirement of the first and the output of the second layer: `layerA >>> layerB` 
+
+### Example: wiring modules together
+Here we define a module to cope with CRUD operations for the `User` domain object. We provide also a live implemntation of the module that depends on a sql connection.
+
+```scala mdoc:silent
+object  userRepo {
+  trait Service {
+    def getUser(userId: UserId): IO[DBError, Option[User]]
+    def createUser(user: User): IO[DBError, Unit]
+  }
+
+  type UserRepo = Has[Service]
+
+  // This simple live version depends only on a DB Connection
+  val inMemory: ZLayer.NoDeps[Nothing, UserRepo] = ZLayer.succeed(
+      new Service {
+        def getUser(userId: UserId): IO[DBError, Option[User]] = ???
+        def createUser(user: User): IO[DBError, Unit] = ???
+      }
+    )
+
+  //accessor methods
+  def getUser(userId: UserId): ZIO[UserRepo, DBError, Option[User]] =
+    ZIO.accessM(_.get.getUser(userId))
+
+  def createUser(user: User): ZIO[UserRepo, DBError, Unit] =
+    ZIO.accessM(_.get.createUser(user))
+}
+```
+
+Then, we define another module to perform some basic logging. The live implementation requires a `logger: Logger`
+
+```scala mdoc:silent
+object logging {
+  trait Service {
+    def info(s: String): UIO[Unit]
+    def error(s: String): UIO[Unit]
+  }
+
+  type Logging = Has[Service]
+
+  val consoleLogger: ZLayer[Console, Nothing, Logging] = ZLayer.fromEnvironment( console =>
```

Comment:
Note some of these names are being updated. This will become `ZLayer#fromFunction` and will automatically lift the value into a `Has` so you don't have to do it.

Reply:
Can you merge upstream changes and make sure this all still compiles? `ZLayer.fromEnvironment` was changed to `ZLayer.fromFunction` and automatically lifts the service into a `Has` for you so you don't have to do that below.

Reply:
I rebased on last master, now it is using `fromFunction`. I'll replace the `(` with `{` as well